File: examples/remote-python/src/worker.py
"""Python worker for the remote-python demo.

Connects to the Node-side Hub on ws://127.0.0.1:4001, registers two agents,
runs until the Hub closes the session.
"""

from __future__ import annotations

import asyncio
import sys

from aipehub import AgentParticipant, connect


class WriterAgent(AgentParticipant):
    def __init__(self) -> None:
        super().__init__(id="writer-py", capabilities=["draft"])

    async def handle_task(self, task: dict) -> dict:
        topic = task["payload"].get("topic", "?")
        print(f"[worker] writer-py received '{task.get('title')}' topic={topic}",
              flush=True)
        await asyncio.sleep(0.3)
        return {
            "text": (
                f"On {topic}: this draft came back over a WebSocket from a Python "
                f"process, and the Hub didn't blink."
            ),
        }


class ReviewerAgent(AgentParticipant):
    def __init__(self) -> None:
        super().__init__(id="reviewer-py", capabilities=["review"])

    async def handle_task(self, task: dict) -> dict:
        print(f"[worker] reviewer-py received '{task.get('title')}'", flush=True)
        await asyncio.sleep(0.3)
        return {"note": "Add a concrete snippet showing the Python SDK in use."}


def _on_state(state: str, info: dict | None) -> None:
    extra = f" ({info})" if info else ""
    print(f"[worker] state -> {state}{extra}", flush=True)


async def main() -> None:
    print("[worker] connecting to ws://127.0.0.1:4001 ...", flush=True)
    session = await connect(
        url="ws://127.0.0.1:4001",
        agents=[WriterAgent(), ReviewerAgent()],
        auto_reconnect=False,
        on_state_change=_on_state,
    )
    print(f"[worker] connected, session_id={session.session_id}", flush=True)
    await session.wait_closed()


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        sys.exit(0)

File: examples/remote-agent/src/worker.ts
/**
 * Worker process: connects to the host's WebSocket transport, registers two
 * remote agents, and runs until the host closes the session.
 */

import { AgentParticipant, connect, type Task } from '@gotong/sdk-node'

class WriterAgent extends AgentParticipant {
  constructor() {
    super({ id: 'writer-remote', capabilities: ['draft'] })
  }
  protected async handleTask(task: Task): Promise<{ text: string }> {
    const payload = task.payload as { topic: string }
    console.log(`[worker] writer-remote received '${task.title}' topic=${payload.topic}`)
    await sleep(300)
    return {
      text: `On ${payload.topic}: it took a WebSocket roundtrip, but here is your sentence.`,
    }
  }
}

class ReviewerAgent extends AgentParticipant {
  constructor() {
    super({ id: 'reviewer-remote', capabilities: ['review'] })
  }
  protected async handleTask(task: Task): Promise<{ note: string }> {
    console.log(`[worker] reviewer-remote received '${task.title}'`)
    await sleep(300)
    return { note: 'Add a concrete code snippet showing the SDK usage.' }
  }
}

async function main(): Promise<void> {
  console.log('[worker] connecting to ws://127.0.0.1:4000 ...')
  const session = await connect({
    url: 'ws://127.0.0.1:4000',
    agents: [new WriterAgent(), new ReviewerAgent()],
    autoReconnect: false,
    onStateChange: (s, info) => {
      console.log(`[worker] state -> ${s}${info?.reason ? ` (${info.reason})` : ''}`)
      if (s === 'closed') process.exit(0)
    },
  })
  console.log(`[worker] connected, sessionId=${session.sessionId}`)

  // Stay alive until the session reaches 'closed' (handled in onStateChange).
  await new Promise<void>(() => {
    /* park forever */
  })
}

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms))
}

main().catch((err) => {
  console.error('[worker] fatal:', err)
  process.exit(1)
})
